File: test/language/packages/cpp/IndexTest.cpp
#include "gtest/gtest.h"
#include "index_workaround/index/Test.h"
#include "zserio/BitStreamReader.h"
#include "zserio/BitStreamWriter.h"

namespace index_workaround
{

class IndexTest : public ::testing::Test
{
protected:
    static constexpr uint32_t ARRAY_SIZE = 10;
};

constexpr uint32_t IndexTest::ARRAY_SIZE;

TEST_F(IndexTest, readWrite)
{
    ::index_workaround::index::Test test;
    test.getIndexes().resize(ARRAY_SIZE);
    test.getIndexesForParameterized().resize(ARRAY_SIZE);
    for (uint32_t i = 0; i < ARRAY_SIZE; ++i)
    {
        test.getArray().emplace_back(i);
        test.getParameterizedArray().emplace_back(i);
    }
    test.initializeChildren();

    zserio::BitBuffer bitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    test.initializeOffsets(writer.getBitPosition());
    test.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ::index_workaround::index::Test readTest(reader);

    ASSERT_EQ(ARRAY_SIZE, readTest.getArray().size());
    ASSERT_EQ(ARRAY_SIZE, readTest.getParameterizedArray().size());
    ASSERT_EQ(test, readTest);
}

} // namespace index_workaround

File: test/language/packages/python/IndexTest.py
import unittest
import zserio

from testutils import getZserioApi

class IndexTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "index.zs")

    def testReadWrite(self):
        test = self.api.Test()
        test.indexes = [ 0 ] * self.ARRAY_SIZE
        test.indexes_for_parameterized = [ 0 ] * self.ARRAY_SIZE
        test.array = [ self.api.Element(i) for i in range(self.ARRAY_SIZE) ]
        test.parameterized_array = [ self.api.ParameterizedElement(i, i) for i in range(self.ARRAY_SIZE) ]

        writer = zserio.BitStreamWriter()
        test.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array)
        readTest = self.api.Test.from_reader(reader)

        self.assertEqual(self.ARRAY_SIZE, len(readTest.array))
        self.assertEqual(self.ARRAY_SIZE, len(readTest.parameterized_array))
        self.assertEqual(test, readTest)

    ARRAY_SIZE = 10
